File: first_missing_possitive.py
# Test Commit 1
def seperateNegatives(arr):
    i=0
    j=0
    l= len(arr)
    for i in range(l):
        if(arr[i]<0):
            temp = arr[i]
            arr[i]=arr[j]
            arr[j] = temp
            j-=-1
    for i in range(j):
        arr=arr[1:]
    return arr


def firstMissingPositive(arr):
    l = len(arr)
    if(l==0):
        return 1
    if(l==1 and arr[0]==1):
        return 2
    elif(l==1):
        return 1
    allN = True
    
    for c in arr:
        if c>0:
            allN = False
            break
    if(allN):
        return 1
    
    arr = seperateNegatives(arr)
    l = len(arr)
    
    for i in range(l):
        ele = abs(arr[i])-1
        if ele<l:
            if(arr[ele]>0):
                arr[ele]*=-1
    # print(arr)

    ans=0
    for c in range(l):
        if(arr[c]>0):
            ans= c+1
            break
    if(ans==0):
        ans = l+1
    return ans
    
    
    

arr=[1,2,3,4,5,6]
# print(arr.index(1))

print(firstMissingPositive(arr))

File: firstMissingPossitive.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    bool allNeg(vector<int> &arr)
    {
        for (int x : arr)
            if (x > 0)
                return false;
        return true;
    }

    int swapNegatives(vector<int> &arr)
    {
        int p = 0;
        for (int i = 0; i < arr.size(); i++)
        {
            if (arr[i] <= 0)
            {
                swap(arr[i], arr[p++]);
            }
        }
        return --p;
    }

    int firstMissingPositive(vector<int> &arr)
    {
        if (arr.size() == 0)
        {
            return 1;
        }
        if (arr.size() == 1)
        {
            if (arr[0] == 1)
            {
                return 2;
            }
            return 1;
        }
        if (allNeg(arr))
        {
            return 1;
        }

        bool oneExist = false;
        for (auto x : arr)
            if (x == 1)
            {
                oneExist = true;
                break;
            }
        if (!oneExist)
        {
            return 1;
        }

        // Now it is asking for min so that will be in range 1 to size+1 THINK!!

        int l = arr.size();
        int negIndex = swapNegatives(arr);

        for (int i = 0; i <= negIndex; i++)
        {
            if (arr[i] == 0)
                arr[i] = 1;
            arr[i] = abs(arr[i]);
        }

        for (int i = negIndex + 1; i < l; i++)
        {
            if (abs(arr[i]) <= l and arr[abs(arr[i]) - 1] > 0)
            {
                int v = arr[abs(arr[i]) - 1];
                arr[abs(arr[i]) - 1] = -1 * v;
            }
        }

        for (int i = 0; i < l; i++)
        {
            if (arr[i] > 0)
            {
                return i + 1;
            }
        }
        return ++l;
    }
};